package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs

```diff
@@ -603,55 +676,120 @@ public class GeometrySubsets {
+    private static Vector3[] SanitizePointsForUnity(Vector3[] vertices, int[] indices)
+    {
+      int indicesCount = indices.Length;
+      Vector3[] sanitizedPoints = new Vector3[indicesCount];
+
+      for (int i = 0; i < indicesCount; i++)
+      {
+        int vertexIndex = indices[i];
+        sanitizedPoints[i] = vertices[vertexIndex];
+      }
+
+      return sanitizedPoints;
+    }
+
+    private static bool IsThereFaceVaryingAttributes(MeshSample usdMesh, SceneImportOptions options)
+    {
+      // If an attribute is face varying, its number of element will be usdMesh.indices.Length.
+      int pointCount = usdMesh.points.Length;
+      int faceCount = usdMesh.faceVertexCounts.Length;
+
+      // Potential face varying attributes are: normals, tangents, color and uv (st, uv, uv2. uv3. uv4).
+      return (IsFaceVaryingAttribute(options.meshOptions.normals, usdMesh.normals, pointCount, faceCount) ||
+        IsFaceVaryingAttribute(options.meshOptions.tangents, usdMesh.tangents, pointCount, faceCount) ||
+        IsFaceVaryingAttribute(options.meshOptions.color, usdMesh.colors, pointCount, faceCount) ||
+        IsFaceVaryingAttribute(options.meshOptions.texcoord0, usdMesh.st) ||
+        IsFaceVaryingAttribute(options.meshOptions.texcoord0, usdMesh.uv) ||
+        options.ShouldBindMaterials /* TODO: Explain why (cf. LoadPrimvars). */);
+    }
+    
+    private static bool IsFaceVaryingAttribute<T>(ImportMode mode, T[] attribute, int pointCount, int faceCount)
+    {
+      // TODO: Attribute interpolation should be retrieve via the USD API, not assumed from the data length.
+      return (
+        ShouldImport(mode) &&
+        attribute != null &&
+        (attribute.Length > pointCount || attribute.Length == faceCount));
+    }
+    
+    private static bool IsFaceVaryingAttribute(ImportMode mode, PrimvarBase attribute)
+    {
+      return (
+        ShouldImport(mode) &&
+        attribute != null &&
+        (
+          attribute.interpolation == PrimvarInterpolation.FaceVarying || 
+          attribute.interpolation == PrimvarInterpolation.Uniform));
+    }
+
+    static void LoadPrimvars(
+      Scene scene,
+      Mesh unityMesh,
+      string usdMeshPath,
+      List<string> primvars,
+      int[] faceVertexCounts,
+      int[] faceVertexIndices)
+    {
+      if (primvars == null || primvars.Count == 0)
+      {
+        return;
+      }
+
       var prim = scene.GetPrimAtPath(usdMeshPath);
-      for (int i = 0; i < primvars.Count; i++) {
+      for (int i = 0; i < primvars.Count; i++)
+      {
         var attr = prim.GetAttribute(new TfToken("primvars:" + primvars[i]));
-        if (!attr) continue;
+        if (!attr)
+        {
+          continue;
+        }
 
         // Read the raw values.
         VtValue val = attr.Get(0.0);
-        if (val.IsEmpty()) { continue; }
+        if (val.IsEmpty())
+        {
+          continue;
+        }
         VtVec2fArray vec2fArray = UsdCs.VtValueToVtVec2fArray(val);
```

Comment:
Oof, I just noticed that we are assuming the primvar is a vec2 here...

Reply:
Should we change that? How?